fix(search) - fixing autocomplete for numeric fields

wmak opened the pull request:
- Currently if a tag is numeric (eg. `stack.lineno`) autocomplete
  doesn't work at all, since we try to perform a `LIKE` or `!=` on an
  int field with a string field
- This changes how we handle numeric keys so that when there's no query
  we return all possible values
- And if there is a query try to find values within 100 of what was
  searched

wmak committed:
fix(search) - fixing autocomplete for numeric tags

- Currently if a tag is numeric (eg. `stack.lineno`) autocomplete
  doesn't work at all, since we try to perform a `LIKE` or `!=` on an
  int field with a string field
- This changes how we handle numeric keys so that when there's no query
  we return all possible values
- And if there is a query try to find values within 100 of what was
  searched

iProgramStuff commented on src/sentry/tagstore/snuba/backend.py:
```diff
@@ -586,13 +587,21 @@ def get_tag_value_paginator_for_projects(
-
-        if query:
-            conditions.append([snuba_key, "LIKE", u"%{}%".format(query)])
+        if key in SearchVisitor.numeric_keys and snuba_key not in BLACKLISTED_COLUMNS:
```
I'm not sure if we'd want this to apply to all numeric keys. Maybe that's the spec though.

I'm thinking for numeric keys like project.id and issue.id, and others that it doesn't make sense to span the condition to be -50, + 50 of the value. Maybe we can define a "fuzzy_numeric_keys" as a list of keys where this kind of search makes sense.

iProgramStuff replied:
I just realized project_id is in BLACKLISTED_COLUMNS. Does project.id get the same treatment? What about issue_id and issue.id?

wmak replied:
:+1: added a list of keys for this search type that excludes both issue/project id.

As far as I can tell this tag value function isn't called for project/issue ids since neither get treated as valid tags by our search bars

iProgramStuff commented on src/sentry/tagstore/snuba/backend.py:
```diff
@@ -586,13 +587,21 @@ def get_tag_value_paginator_for_projects(
+                conditions.append([snuba_key, ">=", converted_query - 50])
+                conditions.append([snuba_key, "<=", converted_query + 50])
+            else:
+                conditions.append([[snuba_key, ">=", 0], [snuba_key, "<", 0]])
```
If the converted_query is None, do we need to add a condition on this field at all? Maybe we can check for not null? If we're just trying to get all results, I think no condition would do the same thing.

wmak replied:
:+1: Good call, looks like no condition does the same thing, will update. 🤔 seems to be the same for the string values too though, I wonder why we did `key != ""`

iProgramStuff reviewed (changes requested):
I tried to run the test but it is failing for me. I wanted to try taking a quick look at how some other fields behave.

wedamija reviewed (commented):
This is really `fixing autocomplete for numeric fields` correct? Things in `numeric_keys` are only fields, there are no tags there.

I'm a little confused about this in general because it looks like it's fetching tags, but you're looking at numeric fields which won't be in tags?

wmak committed:
ref: Addressing some changes from PR review

- Adding constants for key numeric keys, and the distance for those keys
- Tests were written incorrectly (sorry, bad copy pasta)

wedamija reviewed (approved):
Disregard my previous comment, basically just naming confusion. Looks like this method is intended to handle both tags and fields.
